src/accelerate/hooks.py

```diff
@@ -155,17 +155,17 @@ def add_hook_to_module(module: nn.Module, hook: ModelHook, append: bool = False)
     module = hook.init_hook(module)
     module._hf_hook = hook
 
-    @functools.wraps(old_forward)
-    def new_forward(*args, **kwargs):
-        args, kwargs = module._hf_hook.pre_forward(module, *args, **kwargs)
-        if module._hf_hook.no_grad:
+    def new_forward(self, *args, **kwargs):
+        args, kwargs = self._hf_hook.pre_forward(self, *args, **kwargs)
+        if self._hf_hook.no_grad:
             with torch.no_grad():
-                output = old_forward(*args, **kwargs)
+                output = self._old_forward(*args, **kwargs)
         else:
-            output = old_forward(*args, **kwargs)
-        return module._hf_hook.post_forward(module, output)
+            output = self._old_forward(*args, **kwargs)
+        return self._hf_hook.post_forward(self, output)
+
+    module.forward = functools.update_wrapper(functools.partial(new_forward, module), old_forward)
```

Comment:
we can use still use the decorator `@wraps(old_forward)` instead of `update_wrapper` but both are okay to me.

Reply:
If I change to using `@wraps(old_forward)` instead, I end up failing these checks here: https://github.com/huggingface/accelerate/blob/a87c95da9e3b416fb10a0e7dac7d397c015c3ed5/tests/test_hooks.py#L64-L65
This happens since `functools.partial` will happen after `@wraps(forward_function)` occurs, so the signature of the forward function ends up being incorrect, which is why I went with `update_wrapper`.